client library integration

mihajlon97 opened the pull request:
- client library
 - fiskalyServiceUrl optional parameter
 - tests adapted
 - readme updated

Commit:
- client library
 - fiskalyServiceUrl optional parameter
 - tests adapted
 - readme updated

pgaubatz commented on src/client/ClientLibrary.ts:
```diff
@@ -0,0 +1,147 @@
+     */
+    constructor() {
+        try {
+            this.library = ffi.Library(`${__dirname}\\${this.getLibraryName()}`, {
```
this will most likely NOT work on non-windows platforms

mihajlon97 replied:
Is the library path a problem?
Should we use require('path').join()?

pgaubatz replied:
yes, please!

pgaubatz commented on src/FiskalyClient.ts:
```diff
@@ -2,23 +2,33 @@ import * as jayson from 'jayson/promise';
     constructor(fiskalyServiceUrl: string) {
         if (!fiskalyServiceUrl) {
-            throw new FiskalyError("fiskalyServiceUrl must be provided");
+            try {
+                this.clientLibrary = new ClientLibrary();
+                this.hasService = false;
+            } catch (e) {
+                throw new FiskalyError("fiskalyServiceUrl must be provided");
+            }
+        } else {
+            // @ts-ignore
+            this.jsonRPC = jayson.Client.http(fiskalyServiceUrl);
```
how about:
```suggestion
            try {
                const lib = new ClientLibrary();
                this.doRequest = lib.request.bind(lib);
            } catch (e) {
                throw new FiskalyError("fiskalyServiceUrl must be provided");
            }
        } else {
            const rpc = jayson.Client.http(fiskalyServiceUrl);
            this.doRequest = rpc.request.bind(rpc);
```

these class fields can also be removed then:
```ts
    private jsonRPC: jayson.Client | undefined;
    private clientLibrary: ClientLibrary | undefined;
    private hasService: boolean = true;
```

instead there would only be:
```ts
    private doRequest: Function;
```

Commit:
- client library path fix

Commit:
Update src/FiskalyClient.ts

Suggested change -  @pgaubatz

Co-authored-by: Patrick Gaubatz <patrick@fiskaly.com>

Commit:
Merge branch 'client-library' of https://github.com/fiskaly/fiskaly-sdk-node into client-library

Commit:
- suggested change @pgaubatz fix
 - bind request client library method and JsonRpc method into doRequest FiskalyClient method

pgaubatz commented on src/FiskalyClient.ts:
```diff
@@ -2,41 +2,38 @@ import * as jayson from 'jayson/promise';
     private readonly SDK_VERSION = '1.1.600';
     private context: string | undefined;
-    private readonly jsonRPC: jayson.Client;
+    private readonly doRequest: OmitThisParameter<(method: string, params: object) => Promise<JSONRPCResultLike> | null> = () => null;
```
```suggestion
    private readonly doRequestFn: OmitThisParameter<(method: string, params: object) => Promise<JSONRPCResultLike> | null> = () => null;
```

pgaubatz commented on src/FiskalyClient.ts:
```diff
@@ -2,41 +2,38 @@ import * as jayson from 'jayson/promise';
     private context: string | undefined;
-    private readonly jsonRPC: jayson.Client;
+    private readonly doRequest: OmitThisParameter<(method: string, params: object) => Promise<JSONRPCResultLike> | null> = () => null;
+    private jsonRPC: jayson.Client | undefined;
+    private clientLibrary: ClientLibrary | undefined;
+    private hasService: boolean = true;
```
```suggestion
```

pgaubatz commented on src/FiskalyClient.ts:
```diff
@@ -2,41 +2,38 @@ import * as jayson from 'jayson/promise';
-            throw new FiskalyError("fiskalyServiceUrl must be provided");
+            try {
+                const lib = new ClientLibrary();
+                this.doRequest = lib.request.bind(lib);
```
```suggestion
                this.doRequestFn = lib.request.bind(lib);
```

pgaubatz commented on src/FiskalyClient.ts:
```diff
@@ -2,41 +2,38 @@ import * as jayson from 'jayson/promise';
+        } else {
+            // @ts-ignore
+            const rpc = jayson.Client.http(fiskalyServiceUrl);
+            this.doRequest = rpc.request.bind(rpc);
```
```suggestion
            this.doRequestFn = rpc.request.bind(rpc);
```

pgaubatz commented on src/FiskalyClient.ts:
```diff
@@ -62,8 +59,11 @@ export class FiskalyClient {
         const response = await this.doRequest('create-context', contextParams);
-
-        /** Update context */
+        /** Check if error exists */
+        if (response.error != null) {
+            throw FiskalyErrorHandler.throwError(response);
+        }
+        // @ts-ignore
```
```suggestion
```

pgaubatz commented on src/FiskalyClient.ts:
```diff
@@ -89,7 +89,10 @@ export class FiskalyClient {
     public async getVersion(): Promise<VersionResponse> {
         const response = await this.doRequest('version', {});
-
+        /** Check if error exists */
+        if (response.error != null) {
+            throw FiskalyErrorHandler.throwError(response);
+        }
```
```suggestion
```

pgaubatz commented on src/FiskalyClient.ts:
```diff
@@ -105,7 +108,10 @@ export class FiskalyClient {
         };
         const response = await this.doRequest('config', params);
-
+        /** Check if error exists */
+        if (response.error != null) {
+            throw FiskalyErrorHandler.throwError(response);
+        }
```
```suggestion
```

pgaubatz commented on src/FiskalyClient.ts:
```diff
@@ -120,6 +126,10 @@ export class FiskalyClient {
             context: this.context
         };
         const response = await this.doRequest('config', params);
+        /** Check if error exists */
+        if (response.error != null) {
+            throw FiskalyErrorHandler.throwError(response);
+        }
```
```suggestion
```

pgaubatz commented on src/FiskalyClient.ts:
```diff
@@ -138,7 +148,9 @@ export class FiskalyClient {
         };
         const response = await this.doRequest('request', params);
-
+        if (response.error != null) {
+            throw FiskalyErrorHandler.throwError(response);
+        }
```
```suggestion
```

pgaubatz reviewed (changes requested):
Re-add the following method:
```ts
    /**	
     * Send JSON RPC Request to Client	
     * @param {string} method	
     * @param {object} params	
     */	
    private async doRequest(method: string, params: object) {	
        if(!this.allowedMethods.includes(method)) {	
            throw new FiskalyError("Invalid method parameter");	
        }	
        const response = await this.doRequestFn(method, params);	
        /** Check if error exists */	
        if (response.error != null) {	
            throw FiskalyErrorHandler.throwError(response);	
        }	
        return response	
    }
```

Commit:
Update src/FiskalyClient.ts

Co-authored-by: Patrick Gaubatz <patrick@fiskaly.com>

Commit:
Update src/FiskalyClient.ts

Co-authored-by: Patrick Gaubatz <patrick@fiskaly.com>

Commit:
Update src/FiskalyClient.ts

Co-authored-by: Patrick Gaubatz <patrick@fiskaly.com>

Commit:
Update src/FiskalyClient.ts

Co-authored-by: Patrick Gaubatz <patrick@fiskaly.com>

Commit:
Update src/FiskalyClient.ts

Co-authored-by: Patrick Gaubatz <patrick@fiskaly.com>

Commit:
Update src/FiskalyClient.ts

Co-authored-by: Patrick Gaubatz <patrick@fiskaly.com>

Commit:
Update src/FiskalyClient.ts

Co-authored-by: Patrick Gaubatz <patrick@fiskaly.com>

Commit:
Update src/FiskalyClient.ts

Co-authored-by: Patrick Gaubatz <patrick@fiskaly.com>

Commit:
Update src/FiskalyClient.ts

Co-authored-by: Patrick Gaubatz <patrick@fiskaly.com>

Commit:
- doRequestFn method refactoring
 - suggested change @pgaubatz

prempador commented on tests/FiskalyClientTestClientLib.ts:
```diff
@@ -0,0 +1,89 @@
+beforeEach(() => {
+    // @ts-ignore
+    FISKALY_SERVICE_URL = null;
+    FISKALY_API_KEY = 'test_9b92hretadf7nbl9gotci0188_development';
```
Please don't have credentials in code.

mihajlon97 replied:
Fixed now with .env variables.

Commit:
- removed credentials from the tests
 - suggested change @prempador

Commit:
- FISKALY_BASE_URL and FISKALY_SERVICE_URL added to .env too
 - suggested change @prempador

Commit:
Merge branch 'master' into client-library

Commit:
cleanup PR and make it work
